pkg/virt-controller/services/template_test.go

```diff
@@ -4281,6 +4281,59 @@ var _ = Describe("Template", func() {
 			)
 
 		})
+		Context("with configmap in VMI annotations for generic sidecar", func() {
+			var vmi *v1.VirtualMachineInstance
+
+			BeforeEach(func() {
+				vmi = api.NewMinimalVMI("generic-sidecar-test")
+				vmi.Annotations = map[string]string{
+					hooks.HookSidecarListAnnotationName: `[{"image": "test:test", "configMap": {"name": "test-cm", "key": "script.sh"}}]`,
+				}
+			})
+			When("ConfigMap exists on the cluster", func() {
+				BeforeEach(func() {
+					k8sClient := k8sfake.NewSimpleClientset()
+					//virtClient.EXPECT().CoreV1().Return(cm).MaxTimes(1)
+					k8sClient.Fake.PrependReactor("get", "configmaps", func(action testing.Action) (handled bool, obj k8sruntime.Object, err error) {
+						cm := kubev1.ConfigMap{
+							ObjectMeta: metav1.ObjectMeta{
+								Name: "test-cm",
+							},
+							Data: map[string]string{"script.sh": "some-script"},
+						}
+						return true, &cm, nil
+					})
+					virtClient.EXPECT().CoreV1().Return(k8sClient.CoreV1()).AnyTimes()
+				})
+				It("should add ConfigMap as volume to Pod and mount in sidecar", func() {
+					config, kvInformer, svc = configFactory(defaultArch)
+					pod, err := svc.RenderLaunchManifest(vmi)
+					Expect(err).ToNot(HaveOccurred())
+
+					Expect(pod.Spec.Volumes).To(ContainElement(kubev1.Volume{
+						Name: "test-cm",
+						VolumeSource: kubev1.VolumeSource{
+							ConfigMap: &kubev1.ConfigMapVolumeSource{
+								LocalObjectReference: kubev1.LocalObjectReference{Name: "test-cm"},
+								DefaultMode:          pointer.Int32(0777),
+							},
+						},
+					}))
+					Expect(pod.Spec.Containers[1].VolumeMounts).To(ContainElement(kubev1.VolumeMount{
+						MountPath: "/usr/bin/onDefineDomain",
+						Name:      "test-cm",
+						SubPath:   "script.sh",
+					}))
+				})
+			})
+			When("ConfigMap does not exist on the cluster", func() {
```

Comment:
Would it be reasonable to have a webhook to check the VMI requires a ConfigMap and the ConfigMap does not exist?

Reply:
Yeah, it makes sense.

Reply:
Hm, I'm unsure. 

Why can't it just fail in runtime, and keep trying to reconcile ? Meaning, if you later create the config map w/ the required data, you're all set.

Reply:
FWIW, this pattern is what happens currently when you request a network-attachment-definition that doesn't exist.

Reply:
@maiqueb As a user, I would be confused if my VMI didn't come up. A webhook would fail the VMI creation with an error, which is a good UX, IMO.

@victortoso @maiqueb I've never written a webhook myself (but I figured I would have to add it in `vmi-create-admitter.go`), so my only ask is to help me decide which pattern to go with - current or webhook? :)

Reply:
I'd personally go with uniform pattern. 

> @maiqueb As a user, I would be confused if my VMI didn't come up. A webhook would fail the VMI creation with an error, which is a good UX, IMO.

IMHO the next logical step when the user doesn't get what they desire is to `kubectl describe <entity>` . There you would see information stating the dependency for your workload is not met (i.e. the required config map is not available), and you'd create it. Then your workload would be properly started (on a happy path).

The other alternative would require you to clean up the failed resource, then create the dependency, then re-create the resource. More things to do. And a unique flow instead of the same pattern heavily employed in Kubernetes.

Reply:
Thanks for the discussion. I haven't made any change since it complies with the uniform pattern.

> The other alternative would require you to clean up the failed resource, then create the dependency, then re-create the resource. More things to do. And a unique flow instead of the same pattern heavily employed in Kubernetes.

I have a question just to improve/validate my understanding of webhooks.

My understanding is that an admission webhook prevent creation of the resource if a certain condition (here the existence of a configmap) failed? If yes, there won't be a need to clean the failed resource and recreate it, no? Or am I thinking this wrong?

Reply:
You're right. I somehow forgot we were discussing this in the context of webhooks. 

My argument is weaker than what I described. 

Thanks for keeping me honest.

Reply:
Yes, I don't think webhook is a *must* but it is *nice to have*. I personally like the fact that if fails faster then without webhook.

Reply:
While my argument is weaker than described above, a reconciling approach still requires less steps: the user doesn't have to recreate the VMI - it'll keep trying until the missing dependency is added. The user can inspect the vmi object to see why it is not being started.

Anyway ... I won't block it. While we both agree this is not needed, I personally think it doesn't provide a better user experience and requires more validation via a webhook. The latter is something we seem to disagree about, but that's perfectly fine.